app/lib/package/backend.dart

```diff
@@ -909,6 +909,10 @@ class PackageBackend {
         throw InvalidInputException.canonicalizeVersionError(
             pubspec.nonCanonicalVersion);
       }
+      if (versionString != pubspec.nonCanonicalVersion) {
+        throw InvalidInputException.nonCanonicalVersion(
+            pubspec.nonCanonicalVersion, versionString);
+      }
```

Comment:
Why not put this in `summarizePackageArchive` ?

Reply:
I don't think computing the canonical version twice is particularly expensive.

Only concern is whether or not package reader can read existing archives that might violate this property.
But isn't there an explicit `lenient: true` mode to support that? (or do I remember incorrectly?)